diarymanagement-code-r273/dbserver/branches/SSL/dbserver/src/TBA/Messages/ResponseMessage.java

```diff
@@ -0,0 +1,180 @@
+/*
+ * To change this template, choose Tools | Templates
+ * and open the template in the editor.
+ */
+
+package TBA.Messages;
+
+import TBA.Exceptions.MessagesException;
+
+/**
+ * This class is the root of all response messages between the client and
+ * the server. It handles the 'Response Header' section of the message only.
+ *<p>
+ * @author Dan McGrath
+ *
+ * @version $Rev:: 27            $ $Date:: 2009-07-31 #$
+ */
+public class ResponseMessage extends Messages
+{
+   private int errorCode = -1;
+   private String errorMessage = null;
+   private final int errorCodeLen = 1;
+
+   public ResponseMessage() { }
+
+   /**
+    * The Get() Method adds the 'Response Header' to the beginning of the message
+    * passed into it. It then passes it on to Messages to process the
+    * 'Message Header'
+    *<p>
+    * @param childMessage The message to send, not including the 'Message Header'
+    *<p>
+    * @returns The actual message string you can send
+    *<p>
+    * @throws MessagesException
+    *<p>
+    * @see #Set(java.lang.String)
+    */
+   @Override
+   public String Get(String childMessage) throws MessagesException
+   {
+      String message = null;
+
+      if(errorCode == -1)
+      {
+         throw new MessagesException("Error/Success code not set");
+      }
+
+      if(errorCode > 0)
+      {
+         childMessage = errorMessage + '|';
+      }
+
+      if(childMessage == null)
+      {
+         childMessage = "";
+      }
+
+      message = String.format("%0" + Integer.toString(errorCodeLen) + "x", errorCode) + childMessage;
+
+      super.SetMessageType('R');
+      return super.Get(message);
+   }
+
+   /**
+    * The Set() Method takes a Message and strips out the 'Response Header'
+    * information, returning the rest of the message back so that it can then
+    * be processed.
+    *<p>
+    * @param message The message as recieved.
+    *<p>
+    * @returns The message without the 'Response Header'
+    *<p>
+    * @throws MessagesException
+    *<p>
+    * @see #Get(java.lang.String)
+    */
+   @Override
+   public String Set(String message) throws MessagesException
+   { 
+      message = super.Set(message);
+
+      if(message.length() < 1)
+      {
+         throw new MessagesException("The Response Header is incomplete");
+      }
+
+      try
+      {
+         errorCode = Integer.parseInt(message.substring(0,errorCodeLen), 16);
+         message = message.substring(errorCodeLen, message.length());
+      }
+      catch (NumberFormatException ex)
+      {
+         throw new MessagesException("The Error Code is invalid");
+      }
+
+      if(errorCode > 0)
+      {
+         int delimPos = message.indexOf(Character.getNumericValue('|'));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer

If the indexOf() search on line 100 does not find the substring and returns -1, the call to substring() on line 101 will throw IndexOutOfBoundsException.